File: src/main/java/ru/innopolis/interpeter/lexer/tokens/Token.java
package ru.innopolis.interpeter.lexer.tokens;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import ru.innopolis.interpeter.lexer.Code;

@Getter
@AllArgsConstructor
public class Token {
    Span span;
    Code code;
    String rawValue;

    @Override
    public String toString() {
        return String.format("%s(%s)", code, rawValue);
    }
}

File: src/main/scala/ru.innopolis.interpreter/lexer/Token.scala
package ru.innopolis.interpreter.lexer

case class Token[T](span: Span, code: Code, value: T) {
  override def toString: String = s"$code($value)"
}

object Token {
  def apply(code: Code, rawToken: String, span: Span): Token[_] = {
    val base = (value: Any) => Token(span, code, value)

    code match {
      case Code.INT_LITERAL    => base(rawToken.toLong)
      case Code.REAL_LITERAL   => base(rawToken.toDouble)
      case Code.STRING_LITERAL => base(rawToken)
      case Code.IDENTIFIER     => base(rawToken)
      case Code.NEWLINE        => base("\\n")
      case _                   => base(rawToken)
    }
  }
}
